pkg/agent/run.go

```diff
@@ -193,39 +209,67 @@ func gatherData(ctx context.Context, config Config) []*api.DataReading {
 		dataGatherers[dgConfig.Name] = dg
 	}
 
-	// Fetch from all datagatherers
-	now := time.Now()
+	//TODO Change backoff parameters to those desired
+	backOff := backoff.NewExponentialBackOff()
+	backOff.MaxElapsedTime = time.Duration(BackoffMaxTime) * time.Second
 	readings := []*api.DataReading{}
-	failedDataGatherers := []string{}
-	for k, dg := range dataGatherers {
-		i, err := dg.Fetch()
-		if err != nil {
-			log.Printf("Error fetching with DataGatherer %q: %s", k, err)
-			failedDataGatherers = append(failedDataGatherers, k)
-			continue
-		}
+	completedDataGatherers := make(map[string]bool, len(dataGatherers))
 
-		log.Printf("Gathered data for %q:\n", k)
-
-		readings = append(readings, &api.DataReading{
-			ClusterID:    config.ClusterID,
-			DataGatherer: k,
-			Timestamp:    api.Time{Time: now},
-			Data:         i,
-		})
+	// Fetch from all datagatherers
+	getReadings := func() error {
+		var dgError *multierror.Error
+		for k, dg := range dataGatherers {
+			if completedDataGatherers[k] {
+				continue
+			}
+			dgData, err := dg.Fetch()
+			if err != nil {
+				err = fmt.Errorf("%s: %v", k, err)
+				dgError = multierror.Append(dgError, err)
+				continue
+			} else {
+				completedDataGatherers[k] = true
+
+				log.Printf("Successfully gathered data for %q", k)
+				now := time.Now()
+
+				readings = append(readings, &api.DataReading{
+					ClusterID:    config.ClusterID,
+					DataGatherer: k,
+					Timestamp:    api.Time{Time: now},
+					Data:         dgData,
+				})
+			}
+		}
+		dgError.ErrorFormat = func(es []error) string {
+			points := make([]string, len(es))
+			for i, err := range es {
+				points[i] = fmt.Sprintf("* %s", err)
+			}
+			return fmt.Sprintf(
+				"The following %d data gatherer(s) have failed:\n\t%s",
+				len(es), strings.Join(points, "\n\t"))
+		}
+		return dgError
 	}
 
-	if len(failedDataGatherers) > 0 {
-		log.Printf(
-			"Warning, the following DataGatherers failed, %s. Their data is not being sent.",
-			strings.Join(failedDataGatherers, ", "),
-		)
+	err := backoff.RetryNotify(getReadings, backOff, notify)
+	if err != nil {
+		log.Println(err)
+		log.Printf("This will not be retried")
```

Comment:
Definitely not for this change, but it feels like we want to extend the datareadings protocol to be able to include errors for particular gatherers. That would then allow us to expose that to the user in some way.

Reply:
I'm not sure what you mean by this? At the moment the errors are reported for each failed data gatherer.

Reply:
Sorry, I mean reported to the server.

Reply:
Ah I see! That's a good point. So something along the lines of including errors as a reading itself?

Reply:
Yes, something like that. Then the server can tell the difference between "the agent didn't read this" and "the agent wanted to read this and failed" and that will be useful to us in improving the information available to the users about which reports they can view.